chain/evaluator.go

```diff
@@ -0,0 +1,300 @@
+package chain
+
+import (
+	"context"
+
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/go-filecoin/consensus"
+	"github.com/filecoin-project/go-filecoin/metrics"
+	"github.com/filecoin-project/go-filecoin/sampling"
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+var syncOneTimer *metrics.Float64Timer
+
+func init() {
+	syncOneTimer = metrics.NewTimerMs("syncer/sync_one", "Duration of single tipset validation in milliseconds")
+}
+
+// Evaluator defines an interface for evaluating a chain
+type Evaluator interface {
+	Evaluate(context.Context, types.TipSet, []types.TipSet) error
+	IsBadTipSet(types.TipSetKey) bool
+	ExceedsFinalityLimit(ts types.TipSet) bool
+}
+
+type evaluatorChainReaderWriter interface {
+	BlockHeight() (uint64, error)
+	GetHead() types.TipSetKey
+	GetTipSet(tsKey types.TipSetKey) (types.TipSet, error)
+	GetTipSetAndStatesByParentsAndHeight(pTsKey string, h uint64) ([]*TipSetAndState, error)
+	GetTipSetStateRoot(tsKey types.TipSetKey) (cid.Cid, error)
+	HasTipSetAndStatesWithParentsAndHeight(pTsKey string, h uint64) bool
+	PutTipSetAndState(ctx context.Context, tsas *TipSetAndState) error
+	SetHead(ctx context.Context, s types.TipSet) error
+}
+
+// StateEvaluator is a fancy wrapper around calls to expected consensus
+type StateEvaluator struct {
+	// Provides and stores validated tipsets and their state roots.
+	chainStore evaluatorChainReaderWriter
+	// Evaluates tipset messages and stores the resulting states.
+	chainTransitioner syncStateTransitioner
+	// badTipSetCache is used to filter out collections of invalid blocks.
+	badTipSets *badTipSetCache
+}
+
+type syncStateTransitioner interface {
+	// RunStateTransition returns the state root CID resulting from applying the input ts to the
+	// prior `stateRoot`.  It returns an error if the transition is invalid.
+	RunStateTransition(ctx context.Context, ts types.TipSet, ancestors []types.TipSet, stateID cid.Cid) (cid.Cid, error)
+
+	// IsHeaver returns 1 if tipset a is heavier than tipset b and -1 if
+	// tipset b is heavier than tipset a.
+	IsHeavier(ctx context.Context, a, b types.TipSet, aStateID, bStateID cid.Cid) (bool, error)
+}
+
+// NewChainStateEvaluator returns a new StateEvaluator.
+func NewChainStateEvaluator(e syncStateTransitioner, store evaluatorChainReaderWriter) *StateEvaluator {
+	return &StateEvaluator{
+		chainStore:        store,
+		chainTransitioner: e,
+		badTipSets: &badTipSetCache{
+			bad: make(map[string]struct{}),
+		},
+	}
+}
+
+// IsBadTipSet returns true if the tipset `ts` is known to be bad, true otherwise.
+func (cse *StateEvaluator) IsBadTipSet(ts types.TipSetKey) bool {
+	return cse.badTipSets.Has(ts.String())
+}
+
+// Evaluate is a wrapper around calls to consensus, it will either sync the chain `chain` to the store
+// or return an error
+func (cse *StateEvaluator) Evaluate(ctx context.Context, parent types.TipSet, chain []types.TipSet) error {
```

Comment:
`StateEvaluator` is a worse name than `Syncer` from my point of view.  Yes states are evaluated, but this is in order to *validate* tipsets and *sync* the chain store's internal state with the network.